File: Min Absolute Pair Sum/pair.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
using namespace std;
using namespace std::chrono;

int solution(vector<int> &A)
{
    sort(A.begin(), A.end());
    
    auto front = A.end() - 1;
    auto back = A.begin();
    
    int min_sum = 2000000000;
    while (back != front)
    {
        min_sum = min(min_sum, abs(*front + *back));
        if (min_sum == 0) break;

        if (abs(*front) > abs(*back))
            front--;
        else
            back++;
    }

    min_sum = min(min_sum, abs(*front + *back));
    return min_sum;
}

int main()
{
    vector<int> A(100000);
    generate(A.begin(), A.end(), [] {return ((rand() * rand()) % 2000000000) - 1000000000;});
    
    auto start = high_resolution_clock::now();
    cout << solution(A) << endl;
    auto end = high_resolution_clock::now();
    printf("elapsed %.4f seconds", (end - start) * 1e-9);
}

File: Min Absolute Pair Sum/pair.py
def solution(A):
    # write your code in Python 3.6
    A =sorted(A)
    back = 0
    front = len(A) - 1
    
    min_sum = 2000000000
    while(front >= back):
        min_sum = min(min_sum, abs(A[front] + A[back]))
        if min_sum == 0: break
        
        if abs(A[front]) > abs(A[back]):
            front -= 1
        else:
            back += 1

    return min_sum


if __name__ == '__main__':

    import time
    import random

    #A = [-8, 4, 5, -10, 3]
    A = [random.randint(-1000000000, 1000000000) for i in range(0, 100000)]
    
    start = time.time()
    print(solution(A))
    end = time.time()
    print('elapsed %.4f seconds.' % (end - start))
    
